ada/language/ast.py

```diff
@@ -9382,6 +9311,181 @@ class AggregateAssoc(BasicAssoc):
     expr = Property(Entity.r_expr)
     names = Property(Self.designators.map(lambda d: d))
 
+    xref_stop_resolution = Property(True)
+
+    base_aggregate = Property(
+        Entity.parent.parent.cast_or_raise(BaseAggregate)
+    )
+
+    @langkit_property()
+    def xref_equation():
+        agg = Var(Entity.base_aggregate)
+        return Cond(
+            agg.in_aspect('Global'), Entity.globals_assoc_equation,
+
+            agg.in_aspect('Depends'), Entity.depends_assoc_equation,
+
+            agg.parent.is_a(AspectClause, AspectAssoc, PragmaArgumentAssoc),
+            LogicTrue(),
+
+            Entity.general_xref_equation
+        )
+
+    @langkit_property(return_type=Equation, dynamic_vars=[env, origin])
+    def general_xref_equation():
+        """
+        Equation for an aggregate of a record or array.
+        """
+        mra = Var(Entity.base_aggregate.multidim_root_aggregate)
+
+        # If we're part of a multidim aggregate, then take the root aggregate's
+        # type. Else, this is a regular aggregate. In this case grab the type
+        # in type_val.
+        td = Var(If(
+            Not(mra.is_null),
+            mra.typ,
+            Entity.base_aggregate.type_val.cast(BaseTypeDecl),
+        ))
+
+        atd = Var(td.array_def)
+
+        return Cond(
+            atd.is_null,
+            Entity.record_assoc_equation,
+            Entity.array_assoc_equation(atd, mra)
+        )
+
+    @langkit_property(return_type=Equation, dynamic_vars=[env, origin])
+    def record_assoc_equation():
+        """
+        Equation for the case where this is an aggregate assoc for a record
+        type.
+        """
+        agg = Var(Entity.base_aggregate)
+
+        # First, try to find all the discriminants matched by this assoc
+        discr_matches = Var(agg.matched_discriminants.filter(
+            lambda pm: pm.actual.assoc == Entity
+        ))
+
+        # If there are none, this assoc matches one or several components of
+        # the record, so gather them.
+        # WARNING: It is important to gather these components ONLY IF this
+        # association is not for specifying a discriminant. Indeed,
+        # discriminants can (and must) be resolved separately once the type of
```

Comment:
I think the key here is that discriminants must be resolved **first** (not only separately), so that the call to `matched_components` doesn't infinitely recurse (since it needs the value of the discriminants) right ? This is not very clear here.

Reply:
Yes that's right, although the piece of logic here is not for resolving them first, it's really to allow them to be resolved separately from the rest which _enables_ the whole mechanism to resolve them first.